### PR TITLE
Add TTN template

### PR DESCRIPTION
**Pull request by DSGuala**
**Context:**
Adding PennyLane functionality to facilitate the investigation of tensor network based quantum circuits.

**Description of the Change:**
Adds ``qml.templates.tensornetworks.ttn``.
This is a template that creates quantum circuits based on the shape of tree tensor networks.
The tree tensor network template accepts user defined circuits for the individual tensor blocks in the circuit.

**Benefits:**
Allows users to easily build tree tensor network inspired quantum circuits.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: doc/releases/changelog-dev.md
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:7]: Super cool addition!!

File: doc/releases/changelog-dev.md
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:59]:
```suggestion
      qml.TTN(range(n_wires), n_block_wires, block, n_params_block, template_weights)
```

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:84]: If this just contains wire indices, why not use a list or tuple?

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:30]:
```suggestion

    Returns:
```

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:94]: We have used fixed signatures a few times before in PennyLane, and always regretted it :) 

I don't know a better option, but one could at least allow `weights` (or better `parameters`) to be a list of inputs, which gets unpacked as `block(*parameters, wires=...)`. That's how `templates.broadcast` does it...

But it would be great to have a better solution one day!

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:138]: Nice!

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:143]: We removed this redundant class attribute a while back, can you try to delete the line and see what happens?

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:179]: The operator refactor uses a new structure, where hyperparams like `block` are stored in a dictionary called `self.hyperpameters`, so that the set `op.name, op.wires, op.parameters, op.hyperparameters` fully describes an op (in the sense that if these are equal, two ops can be considered equal.

Don't worry about it for now, we'll adapt the template (unless you want to give it a go)! But things will be a bit easier then.

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [inline pennylane/templates/tensornetworks/ttn.py:181]: In the same refactor we start defining decompositions instead of expand, but I'll take care of this!

File: tests/templates/test_tensornetworks/test_TTN.py
Comment (mariaschuld) [inline tests/templates/test_tensornetworks/test_TTN.py:316]: Question: do you guarantee that the template is trainable wrt the block params? I think it should be, but for that you'd have to add diffability tests...

**Review by mariaschuld — COMMENTED** [review]
Only minor comments for now, looks good. We will redesign it a bit in the op refactor!

File: pennylane/templates/tensornetworks/ttn.py
Comment (josh146) [inline pennylane/templates/tensornetworks/ttn.py:134]: @DSGuala if you set `grad_method=None` below, this should register that this operation has no explicit gradient recipe, and should automatically be decomposed. This should mean that `expansion_strategy='device'` is not needed here, I think.

File: pennylane/templates/tensornetworks/ttn.py
Comment (ixfoduap) [reply pennylane/templates/tensornetworks/ttn.py:32]: Yes, I believe it's wire labels, but @DSGuala can confirm

File: pennylane/templates/tensornetworks/ttn.py
Comment (ixfoduap) [inline pennylane/templates/tensornetworks/ttn.py:51]: Could this ever throw an error due to numerical precision? I guess not if `n_wires` and `n_block_wires` are integers, but good to check

File: pennylane/templates/tensornetworks/ttn.py
Comment (ixfoduap) [inline pennylane/templates/tensornetworks/ttn.py:56]: I'm confused. Seems like you are computing `2^(log2(n/m)) = n/m`. So why not just `n/m`? Am I missing something?

File: doc/introduction/templates.rst
Comment (ixfoduap) [inline doc/introduction/templates.rst:190]: The image woud look better and convey the idea of the template better if you used blocks of two qubits instead of blocks of four. It's more work, I know, but worth it

File: pennylane/templates/tensornetworks/ttn.py
Comment (ixfoduap) [inline pennylane/templates/tensornetworks/ttn.py:161]: Same here, seems like we can just use `2^\log_2(x) = x`

File: doc/introduction/templates.rst
Comment (ixfoduap) [inline doc/introduction/templates.rst:189]:
```suggestion
    :description: Tree Tensor Network
```

File: doc/introduction/templates.rst
Comment (ixfoduap) [inline doc/introduction/templates.rst:184]:
```suggestion
    :description: Matrix Product State
```

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Looking really good!

@mariaschuld Has made great suggestions, specially regarding differentiability tests. Please address those.

Other than that, my only major suggestion is to update the image to better showcase the tree structure of the template

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:32]: Indeed this is wire lables. I'll make the change.

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:51]: We can get a false `True` when n_block_wires is e.g. `4.000000000000001`.
I think users will feed integers to the template. This should work both when the arguments are of type `float` or of type `int`.
I could raise an error or warning when `n_wires` or `n_block_wires` are not integers.

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:56]: `2^int(log2(n/m)) = 2^floor(log2(n/m)) <= n/m`
This takes care of the case when we have e.g. 3 wires, 2 wires per block so we want to use only the first 2 wires.
Then `n/m = 1.5` but `2^int(log2(n/m))=2`
At first glance, it will look like we can use something simpler, like `n//m*m`. But for cases like 13 wires, 4 wires per block, we get `n//m*m = 13//4*4 = 12`. Here, `12` is will not give a viable TTN. For a TTN of block size 4, we can use `[4, 8, 16, 32, etc.]` wires. The next step up would be 16 wires. So we have to use the logarithm.
We can discuss this more if it's not clear.

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:85]: Thanks! I'll make this change.

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:133]: I made this change but still seem to require `expansion_strategy='device'` :thinking:

Comment (josh146) [reply pennylane/templates/tensornetworks/ttn.py:133]: oh 🤔 Oh! this is probably because the QNode is in backprop mode; I think removing the `expansion_strategy` would only work in parameter-shift mode.

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:93]: Thanks Maria! I made this change. I'm happy that it makes `block` more flexible.
I'm now getting some warnings about ragged arrays. I think somewhere deeper in PennyLane the parameters are converted to numpy arrays :thinking:

File: tests/templates/test_tensornetworks/test_TTN.py
Comment (DSGuala) [reply tests/templates/test_tensornetworks/test_TTN.py:399]: I added a differentiability test. This seems to work fine except for ragged arrays. I suppose this is like the warnings in the other comment and related to how PennyLane handles the parameters :thinking:

File: pennylane/templates/tensornetworks/ttn.py
Comment (mariaschuld) [reply pennylane/templates/tensornetworks/ttn.py:93]: That is very strange, because we should only ever work with lists of tensors/arrays, or with flattened versions. I think it would be good to trace where this error arises, just in case it is something trivial. Ideally one should never create a ragged array, they will be increasingly discontinued...

File: pennylane/templates/tensornetworks/ttn.py
Comment (ixfoduap) [reply pennylane/templates/tensornetworks/ttn.py:56]: Got it, thanks!

File: pennylane/templates/tensornetworks/ttn.py
Comment (ixfoduap) [reply pennylane/templates/tensornetworks/ttn.py:51]: No it's fine as is, thanks!

**Review by ixfoduap — APPROVED** [review]
Looks good from my side, but let's wait for an approval from @mariaschuld before merging

File: pennylane/templates/tensornetworks/ttn.py
Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:93]: I found that this happens with `qml.broadcast()` as well:
```
import pennylane as qml

def mytemplate(par1, par2, wires):
    qml.RY(par1[0], wires=wires[0])
    qml.RY(par1[1], wires=wires[0])
    qml.RY(par2,wires=wires[1])

dev = qml.device('default.qubit', wires=4)

@qml.qnode(dev)
def circuit(pars):
    qml.broadcast(unitary=mytemplate, pattern="chain", wires=[0,1,2,3], parameters=pars)
    return qml.expval(qml.PauliZ(0))

print(qml.grad(circuit)([[[1.0,1.0],2.0],[[1.0,1.0],2.0],[[1.0,1.0],2.0]]))
```
Error: array is not broadcastable to correct shape

Comment (DSGuala) [reply pennylane/templates/tensornetworks/ttn.py:93]: I traced the ragged array warning to the `construct()` method of  `qnode.py` :

```
def construct(self, args, kwargs):
        """Call the quantum function with a tape context, ensuring the operations get queued."""

        if self.interface == "autograd":
            # HOTFIX: to maintain backwards compatibility existing PennyLane code and demos, here we treat
            # all inputs that do not explicitly specify `requires_grad=False`
            # as trainable. This should be removed at some point, forcing users
            # to specify `requires_grad=True` for trainable parameters.
            args = [
                qml.numpy.array(a, requires_grad=True) if not hasattr(a, "requires_grad") else a
                for a in args
            ]
...
```
Here args is a qml.numpy.array made from `args`. `args` only has one element,`a`. This `a` is the ragged array supplied by a user.

I think it may be okay to leave things like this and put a warning in the documentation about the use of ragged arrays to define the template parameters :thinking:

**Review by mariaschuld — APPROVED** [review]
I didn't check every little detail, but looks good overall!

File: pennylane/templates/tensornetworks/ttn.py
Comment (josh146) [reply pennylane/templates/tensornetworks/ttn.py]: @DSGuala luckily we are planning to remove that HOTFIX if statement within days/weeks :)

Comment (josh146) [reply pennylane/templates/tensornetworks/ttn.py]: @antalszava tagging you here for visibility

**Pull request merged**
